File: 20-valid-parentheses/20-valid-parentheses.java
//Simple linear solution to just check all of them.
//Question: Is there a cleaner solution?

class Solution {
    public boolean isValid(String s) {
        Stack<Character> openStack = new Stack<>();
        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[') {
                openStack.push(s.charAt(i));
            } else if (openStack.isEmpty()) {
                return false;  
            } else if (s.charAt(i) == ')') {
                if(openStack.peek() != '(') return false;
                openStack.pop();
            } else if (s.charAt(i) == '}') {
                if(openStack.peek() != '{') return false;
                openStack.pop();
            } else if (s.charAt(i) == ']') {
                if(openStack.peek() != '[') return false;
                openStack.pop();
            }
        }
        return openStack.isEmpty();
    }
}

File: 20-valid-parentheses/20-valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        //Approach: Use a stack, push open and pop closed paren
        val stack = mutableListOf<Char>()
        s.forEach{ char ->
            when (char) {
                '(' , '{' , '[' -> stack.add(0, char) //Add to front
                ')' -> if(stack.elementAtOrNull(0) != '(') return false else stack.removeAt(0)
                ']' -> if(stack.elementAtOrNull(0) != '[') return false else stack.removeAt(0)
                '}' -> if(stack.elementAtOrNull(0) != '{') return false else stack.removeAt(0)
            }
        }
        return stack.isEmpty()
    }
}